stellargraph/data/explorer.py

```diff
@@ -451,24 +435,26 @@ def transition_probability(nn):
                 current_node = node
 
                 for _ in range(length - 1):
-                    neighbours = self.graph.neighbors(
-                        current_node, use_ilocs=True, include_edge_weight=weighted
+                    # select one of the neighbours using the
+                    # appropriate transition probabilities
+                    neighbours, weights = self.graph.neighbor_arrays(
+                        current_node, include_edge_weight=True, use_ilocs=True
                     )
 
-                    if not neighbours:
+                    if len(neighbours) == 0:
                         break
 
-                    # select one of the neighbours using the
-                    # appropriate transition probabilities
-                    choice = naive_weighted_choices(
-                        rs, (transition_probability(nn) for nn in neighbours),
-                    )
+                    mask = neighbours == previous_node
+                    weights[mask] *= ip
+                    mask |= np.isin(neighbours, previous_node_neighbours)
+                    weights[~mask] *= iq
+
+                    probs = np.cumsum(weights) / weights.sum()
+                    choice = np.where(probs > rs.random())[0][0]
```

Comment:
I think this can use http://docs.scipy.org/doc/numpy/reference/generated/numpy.searchsorted.html:

```suggestion
                    choice = np.searchsorted(probs, rs.random(), side="right")
```

Base on https://stackoverflow.com/a/25032853/1256624 this is significantly faster for large arrays, and only fractionally slower for small arrays:

| N | where | searchsorted |
|---|---|---|
| 3 | 2.1 | 2.7 |
| 10 | 2.2 | 2.6 |
| 100 | 2.5 | 2.8 |
| 1000 | 5.3 | 3.7 |
| 10000 | 26.3  | 11.5 |

Measured with:

```python
import numpy as np
N = 10000
aa = np.arange(-N,N)
%timeit -n 100000 np.searchsorted(aa, N/2)+1
%timeit -n 100000 np.where(aa>N/2)[0][0]
```

In particular, this is trading off a little bit of speed for better handling of "super-vertices" with many connections. What do you think?

Reply:
oh that's a good idea! I think a random walk is more likely to end up at super-vetex so I reckon this trade-off is worth it :+!: